reinforcement_learning/bandits_recsys_movielens_testbed/src/env.py

```diff
@@ -0,0 +1,144 @@
+        if self.step_count > self.total_users - 1:
+            self.step_count = 0
+
+        if self.max_users:
+            if self.step_count > self.max_users - 1:
+                self.step_count = 0
+
+        # TODO: Randomize user selection
+        self.current_user_id = self.step_count
+        self.current_user_embedding = None
+
+        # List of all the items that the user has rated in the past
+        self.current_item_pool = np.flatnonzero(self.attractiveness_means[self.current_user_id])
+        if self.item_pool_size and (len(self.current_item_pool) > self.item_pool_size):
+            random_indices = np.random.choice(len(self.current_item_pool), size=self.item_pool_size, replace=False)
+            self.current_item_pool = self.current_item_pool[random_indices]
+        self.current_items_embedding = self.item_features[self.current_item_pool]
+
+    def step(self, actions):
+        assert len(actions) == self.top_k, "Size of recommended items list does not match top-k"
+        rewards, regret, random_regret = self.get_feedback(actions)
+        self.total_regret += regret
+        self.total_random_regret += random_regret
+        info = {"total_regret": self.total_regret, "total_random_regret": self.total_random_regret}
+        self.step_count += 1
+        self._regulate_item_pool()
+        return (self.current_user_embedding, self.current_items_embedding), rewards, False, info
+
+    def get_feedback(self, actions, click_model="cascade"):
+        """
+        Return rewards: List[float] and regret for the current recommended list - actions
+
+        Args:
+            actions: A list of top-k actions indices picked by the agent from candidate list
+            click_model: One of 'cascade', 'pbm'
+
+        Returns:
+            rewards: A reward corresponding to each item in the list
+            regret: Expected regret calculated based on the recommended actions
+            regret_random: Expected regret calculated based on the actions of a random agent
+        """
+        # TODO: Implement PBM: Position based model
+
+        recommended_item_ids = self.current_item_pool[actions]
+        attraction_probs = self.attractiveness_means[self.step_count][recommended_item_ids]
+        
+        random_indices = np.random.choice(len(recommended_item_ids), size=self.top_k, replace=False)
+        random_item_ids = self.current_item_pool[random_indices]
+        random_attraction_probs = self.attractiveness_means[self.step_count][random_item_ids]
+        # Simulate user behavior using a cascading click model.
+        # User scans the list top-down and clicks on an item with prob = attractiveness_means.
+        # User stops seeing the list after the first click.
+
+        clicks = np.random.binomial(1, attraction_probs)
+        if clicks.sum() > 1:
+            first_click = np.flatnonzero(clicks)[0]
+            clicks = clicks[:first_click + 1]
+
+        expected_reward = 1 - np.prod(1 - attraction_probs)
```

Comment:
should the expected reward be mean of a negative binomial distribution?

Reply:
Looking into it. Will compare the plots as well.